src/cargo/ops/cargo_output_metadata.rs

```diff
@@ -115,27 +115,43 @@ fn build_resolve_graph(
     let target_data = RustcTargetData::new(ws, &requested_kinds)?;
     // Resolve entire workspace.
     let specs = Packages::All.to_package_id_specs(ws)?;
-    let resolve_opts = ResolveOpts::new(
-        /*dev_deps*/ true,
+    let requested_features = RequestedFeatures::from_command_line(
         &metadata_opts.features,
         metadata_opts.all_features,
         !metadata_opts.no_default_features,
     );
+    let resolve_opts = ResolveOpts::new(/*dev_deps*/ true, requested_features.clone());
+    let force_all = if metadata_opts.filter_platforms.is_empty() {
+        crate::core::resolver::features::ForceAllTargets::Yes
+    } else {
+        crate::core::resolver::features::ForceAllTargets::No
+    };
     let ws_resolve = ops::resolve_ws_with_opts(
         ws,
         &target_data,
         &requested_kinds,
         &resolve_opts,
         &specs,
         HasDevUnits::Yes,
-        crate::core::resolver::features::ForceAllTargets::No,
+        force_all,
     )?;
-    // Download all Packages. This is needed to serialize the information
-    // for every package. In theory this could honor target filtering,
-    // but that would be somewhat complex.
+
+    // Download all Packages. This is needed to serialize the information for every package.
+    // Note that even with --filter-platform we end up downloading host dependencies as well,
+    // as that is the behavior of download_accessible.
     let package_map: BTreeMap<PackageId, Package> = ws_resolve
         .pkg_set
-        .get_many(ws_resolve.pkg_set.package_ids())?
+        .download_accessible(
+            &ws_resolve.targeted_resolve,
+            &ws.members_with_features(&specs, &requested_features)?
+                .into_iter()
+                .map(|(p, _)| p.package_id())
+                .collect::<Vec<_>>(),
+            HasDevUnits::Yes,
+            &requested_kinds,
+            &target_data,
+            force_all,
+        )?
```

Comment:
I don't think this is necessary, is it?  I think everything should have been downloaded in `resolve_ws_with_opts`.

Perhaps it can be replaced with a new PackageSet method like this:

```rust
pub fn packages(&self) -> impl Iterator<Item = &Package> {
    self.packages.values().filter_map(|p| p.borrow())
}
```

Reply:
Oh, that's a great observation! I did not realize `resolve_ws_with_opts` did downloads as well -- will update shortly. I wonder why the old code did `get_many` here then?